### PR TITLE
Add API endpoint settings for Android

### PR DESCRIPTION
**Pull request by dlon**
The primary purpose of this PR is to make it easier to mock/test the API. Main changes:

* Add `MULLVAD_API_DISABLE_TLS` option.
* Disable bridges when the API IP is overridden.
* Resolve the API IP from hostname if `MULLVAD_API_HOST` is set but not `MULLVAD_API_ADDR`.
* Add API endpoint settings for Android.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153)

### CODE REVIEW

File: mullvad-api/src/lib.rs [hunk 1]
```diff
@@ -96,27 +132,45 @@ impl ApiEndpoint {
...
-                        .parse()
-                        .expect("MULLVAD_API_ADDR is not a valid socketaddr");
-                    api.disable_address_cache = true;
-                    log::debug!("Overriding API. Using {} at {}", api.host, api.addr);
-                }
+        if cfg!(feature = "api-override") && (host_var.is_some() || address_var.is_some()) {
+            use std::net::ToSocketAddrs;
+
+            let scheme;
+            if let Some(tls_var) = disable_tls_var {
```
Comment (faern) [inline mullvad-api/src/lib.rs:149]: Nit. But you can do `let scheme = if let....`. I think this has a slight better quality to it, as it becomes more apparent exactly where it's initialized.

File: mullvad-api/src/lib.rs [hunk 2]
```diff
@@ -96,27 +132,45 @@ impl ApiEndpoint {
...
+                api.disable_tls = tls_var != "0";
```
Comment (faern) [inline mullvad-api/src/lib.rs:150]: When shortened from `disable_tls_var` to just `tls_var` I think important information got lost. It's critical to keep the logic correct here, but it's not obvious what `tls_var != "0"` should mean really. You can reuse the same name here (`if let Some(disable_tls_var) = disable_tls_var ...`)

**Review by faern — COMMENTED** [review]
I know the PR is just WIP. But I looked over it and left some comments.

File: mullvad-api/src/lib.rs [hunk 3]
```diff
@@ -96,27 +132,45 @@ impl ApiEndpoint {
...
+                api.addr = format!("{}:{}", api.host, API_PORT_DEFAULT)
+                    .to_socket_addrs()
+                    .expect("failed to resolve API host")
+                    .next()
+                    .expect("API host yielded 0 addresses");
             }
+            api.disable_address_cache = true;
+            api.force_direct_connection = true;
+            log::debug!("Overriding API. Using {} at {scheme}{}", api.host, api.addr);
         } else if host_var.is_some() || address_var.is_some() {
```
[the remark this answers is not in the document]
Comment (dlon) [reply mullvad-api/src/lib.rs]: Done.

File: mullvad-api/src/lib.rs (hunk 1 above)
Comment (dlon) [reply mullvad-api/src/lib.rs]: Done.

**Review by albin-mullvad — COMMENTED** [review]
The Android changes :lgtm: 

> Reviewed 1 of 11 files at r1, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153#-:-NHUJnA3CmauKuuh-cnN:b4r9zw2)* status: 3 of 11 files reviewed, 2 unresolved discussions (waiting on @faern)

**Review by pinkisemils — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153)* status: 3 of 11 files reviewed, 5 unresolved discussions (waiting on @dlon and @faern)

___
*[`mullvad-api/src/https_client_with_sni.rs` line 311 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153#-NHYoTjV9M23lW1XWzvY:-NHYoTjV9M23lW1XWzvZ:bfadcw0) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6067ca14d12310734ba336d8a8c1dd4179aa0b77/mullvad-api/src/https_client_with_sni.rs#L311)):*
> ```XML
>                             )
>                             .await?;
>                             if API.disable_tls {
> ```

Personally, I'd greatly appreciate if this environment variable would only ever be settable if a specific feature was enabled. 
___
*[`mullvad-api/src/lib.rs` line 109 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153#-NHYpX4X2vI2ckjiXcWL:-NHYpX4X2vI2ckjiXcWM:b-qrya4w) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6067ca14d12310734ba336d8a8c1dd4179aa0b77/mullvad-api/src/lib.rs#L109)):*
> ```XML
>     pub disable_address_cache: bool,
>     pub disable_tls: bool,
>     pub force_direct_connection: bool,
> ```

I think these two fields shouldn't exist in a production build. and could be gated with the feature `api-override`. 
___
*[`mullvad-api/src/proxy.rs` line 167 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153#-NHYqsAs5ZysEq_uJocQ:-NHYqsAs5ZysEq_uJocR:b-wzj8mu) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6067ca14d12310734ba336d8a8c1dd4179aa0b77/mullvad-api/src/proxy.rs#L167)):*
> ```XML
> }
> 
> trait AnyConnection: AsyncRead + AsyncWrite + Unpin + Connection + Send {}
> ```

`AnyConnection` is somewhat confusing - I think it'd be better to call it `hyper::something::Connection` and refer to the imported connection by using (part of) it's import path.

**Review by dlon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153)* status: 3 of 13 files reviewed, 4 unresolved discussions (waiting on @faern and @pinkisemils)

___
*[`mullvad-api/src/lib.rs` line 109 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153#-NHYpX4X2vI2ckjiXcWL:-NH_6YEgCV8q1sN0HBJK:b-mrpgw6) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6067ca14d12310734ba336d8a8c1dd4179aa0b77/mullvad-api/src/lib.rs#L109)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

I think these two fields shouldn't exist in a production build. and could be gated with the feature `api-override`. 
</blockquote></details>

Done. But would it be acceptable to make everything read-only instead? I suspect the increased readability would justify that.
___
*[`mullvad-api/src/proxy.rs` line 167 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153#-NHYqsAs5ZysEq_uJocQ:-NH_8uffANXJU6T-Lo25:b5qxfmb) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6067ca14d12310734ba336d8a8c1dd4179aa0b77/mullvad-api/src/proxy.rs#L167)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

`AnyConnection` is somewhat confusing - I think it'd be better to call it `hyper::something::Connection` and refer to the imported connection by using (part of) it's import path.
</blockquote></details>

Is this what you had in mind?

**Review by pinkisemils — COMMENTED** [review]
> Reviewed 2 of 7 files at r5.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153)* status: 5 of 13 files reviewed, 3 unresolved discussions (waiting on @dlon and @faern)

___
*[`mullvad-api/src/lib.rs` line 109 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153#-NHYpX4X2vI2ckjiXcWL:-NHeC1Lk5zKiPShktSM1:b7tza5l) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6067ca14d12310734ba336d8a8c1dd4179aa0b77/mullvad-api/src/lib.rs#L109)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

Done. But would it be acceptable to make everything read-only instead? I suspect the increased readability would justify that.
</blockquote></details>

I think for readability, the extra members could be put in an `ApiEndpointOverride` struct. Or maybe this structure should be an enum instead, with variants that would override regular endpoints only existing if the feature is turned on?
___
*[`mullvad-api/src/proxy.rs` line 167 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153#-NHYqsAs5ZysEq_uJocQ:-NHeBp9R84EVAKvpLTxd:b3vs28z) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6067ca14d12310734ba336d8a8c1dd4179aa0b77/mullvad-api/src/proxy.rs#L167)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

Is this what you had in mind?
</blockquote></details>

Yep. :lgtm:

**Review by pinkisemils — COMMENTED** [review]
> Reviewed 6 of 11 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 5 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153)* status: all files reviewed, 3 unresolved discussions (waiting on @dlon and @faern)

**Review by pinkisemils — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4153)* status: all files reviewed, 2 unresolved discussions (waiting on @faern)

**Pull request merged**